allensdk/test/api/test_caching_utilities.py

```diff
@@ -0,0 +1,263 @@
+from functools import partial
+import re
+import os
+
+import pytest
+import pandas as pd
+
+from allensdk.api import caching_utilities as cu
+
+
+def get_data():
+    return pd.DataFrame(
+        {"a": [1, 2, 3, 4], "b": ["duck", "kangaroo", "walrus", "ibex"]}
+    )
+
+
+def swapped_data():
+    return pd.DataFrame(
+        {"b": [1, 2, 3, 4], "a": ["duck", "kangaroo", "walrus", "ibex"]}
+    )
+
+
+def write_to_dict(dc, data):
+    dc["data"] = data
+
+
+def read_from_dict(dc):
+    return dc["data"]
+
+
+class InitiallyFailing:
+    def __init__(self, succeed_at):
+        self.succeed_at = succeed_at
+        self.count = 0
+
+    def __call__(self, fn):
+        self.count += 1
+
+        if self.count >= self.succeed_at:
+            return fn()
+        else:
+            raise ValueError("foo!")
+
+
+class InitiallyFailingWriter(InitiallyFailing):
+    def __call__(self, dc, data):
+        super(InitiallyFailingWriter, self).__call__(partial(write_to_dict, dc, data))
+
+
+class InitiallyFailingReader(InitiallyFailing):
+    def __call__(self, dc):
+        return super(InitiallyFailingReader, self).__call__(partial(read_from_dict, dc))
+
+
+class CallCountingCleanup:
+    def __init__(self):
+        self.count = 0
+
+    def __call__(self, dc):
+        self.count += 1
+        dc.pop("data", None)
+
+
+class CallCountingFetch:
+    def __init__(self):
+        self.count = 0
+
+    def __call__(self):
+        self.count += 1
+        return get_data()
+
+
+def swap(data):
+    data = data.copy()
+    tmp = data["a"]
+    data["a"] = data["b"]
+    data["b"] = tmp
+    return data
+
+
+@pytest.mark.parametrize(
+    "existing,fetch,write,read,pre_write,cleanup,lazy,num_tries,failure_message,expected,expected_fetches,expected_cleanups",
+    [
+        pytest.param(
+            False,
+            CallCountingFetch(),
+            write_to_dict,
+            InitiallyFailingReader(2),
+            swap,
+            CallCountingCleanup(),
+            True,
+            1,
+            "",
+            swapped_data(),
+            1,
+            0,
+            id="simple case"
```

Comment:
Maybe these test cases would be more readable if you stored them as dicts?

```
{
 "existing": False,
 "fetch": CallCountingFetch(),
 "write": write_to_dict,
 "read": InitiallyFailingReader(2),
 "pre_write": swap,
 "cleanup": CallCountingCleanup(),
 "lazy": True
 etc...
}
```

Reply:
if that works with pytest natively It sounds much better than the way I've got it. 

The tall formatting is a result of running black. I kind of hate it, but with dicts it would not be so bad.

Reply:
it does not work with pytest natively. Alternative is to have 1 param, which is a dict and pop off values in the test. Not sure if that is better.